File: week5/lee/정수삼각형.py
def solution(triangle):
    answer = 0
    dp = [[0 for j in range(i+1)] for i in range(len(triangle))]
    # print(dp)
    for idx,i in enumerate(dp):
        if idx == 0:# 맨처음은 그냥 넣어주기
            dp[idx][0] = triangle[0][0]
            continue

        for idxj, j in enumerate(i):
            if idxj == 0:#처음
                dp[idx][idxj] = dp[idx-1][0]+triangle[idx][idxj]
            elif idxj == len(i)-1:# 마지막
                dp[idx][idxj] = dp[idx-1][idxj-1]+triangle[idx][idxj]
            else:
                dp[idx][idxj] = max(dp[idx-1][idxj-1], dp[idx-1][idxj]) + triangle[idx][idxj]


    return max(dp[-1])

print(solution([[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]]	))

File: week5/eunbeen/정수삼각형.java
import java.util.*;

class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;

        int length = triangle.length;

        int[][] dp = new int[length][length];
        dp[0][0] = triangle[0][0];
        dp[1][0] = dp[0][0] + triangle[1][0];
        dp[1][1] = dp[0][0] + triangle[1][1];
        
        for(int i=2; i<length; i++){
            dp[i][0] = triangle[i][0] + dp[i-1][0];
            for(int j=1; j<i; j++){
                dp[i][j] = triangle[i][j] + Math.max(dp[i-1][j-1], dp[i-1][j]);
            }
            dp[i][i] = triangle[i][i] + dp[i-1][i-1];
        }
        
        for(int item : dp[length-1]){
            answer = Math.max(item, answer);
        }
        
        return answer;
    }
}


/*
7
3 8
8 1 0
2 7 4 4 
4 5 2 6 5



*/